provider/bigquery.go

```diff
@@ -0,0 +1,1382 @@
+
+	bqQ := pt.client.Query(query)
+
+	it, err := bqQ.Read(pt.query.getContext())
+	if err != nil {
+		return 0, err
+	}
+
+	err = it.Next(&n)
+	if err != nil {
+		return 0, err
+	}
+
+	return n[0].(int64), nil
+}
+
+func (pt *bqPrimaryTable) Write(rec GenericRecord) error {
+	tb := pt.name
+	recordsParameter, columns, columnsString := pt.getNonNullRecords(rec)
+
+	placeholder := pt.query.createValuePlaceholderString(columns)
+	upsertQuery := pt.query.upsertQuery(tb, columnsString, placeholder)
+
+	bqQ := pt.client.Query(upsertQuery)
+	bqQ.Parameters = recordsParameter
+
+	_, err := bqQ.Read(pt.query.getContext())
+
+	return err
+}
+
+func (pt *bqPrimaryTable) getNonNullRecords(rec GenericRecord) ([]bigquery.QueryParameter, []TableColumn, string) {
+	recordsParameter := make([]bigquery.QueryParameter, 0)
+	recordColumns := make([]TableColumn, 0)
+	selectedColumns := make([]string, 0)
+	for i, r := range rec {
+		if r == nil {
+			continue
+		}
+		recordsParameter = append(recordsParameter, bigquery.QueryParameter{Value: r})
+		recordColumns = append(recordColumns, pt.schema.Columns[i])
+		selectedColumns = append(selectedColumns, pt.schema.Columns[i].Name)
+	}
+
+	columns := strings.Join(selectedColumns, ",")
+
+	return recordsParameter, recordColumns, columns
+}
+
+func (it *bqGenericTableIterator) Next() bool {
+	var rowValues []bigquery.Value
+	err := it.iter.Next(&rowValues)
+	if err == iterator.Done {
+		it.err = nil
+		return false
+	} else if err != nil {
+		it.err = err
+		return false
+	}
+
+	genRows := make(GenericRecord, len(rowValues))
+	for i, value := range rowValues {
+		if value == nil {
+			continue
+		}
+
+		colType := it.iter.Schema[i].Type
+		genRows[i] = it.query.castTableItemType(value, colType)
+	}
+	it.currentValue = genRows
+	return true
+}
+
+func (it *bqGenericTableIterator) Values() GenericRecord {
+	return it.currentValue
+}
+
+func (it *bqGenericTableIterator) Columns() []string {
+	var columns []string
+	for _, col := range it.iter.Schema {
+		columns = append(columns, col.Name)
+	}
+	return columns
+}
+
+func (it *bqGenericTableIterator) Err() error {
+	return it.err
+}
+
+func newBigQueryGenericTableIterator(it *bigquery.RowIterator, query BQOfflineTableQueries) GenericTableIterator {
+	return &bqGenericTableIterator{
+		iter:         it,
+		currentValue: nil,
+		err:          nil,
+		query:        query,
+	}
+}
+
+func (q defaultBQQueries) registerResources(client *bigquery.Client, tableName string, schema ResourceSchema, timestamp bool) error {
+	var query string
+	if timestamp {
+		query = fmt.Sprintf("CREATE VIEW `%s` AS SELECT %s as entity, %s as value, %s as ts FROM `%s`", q.getTableName(tableName),
+			schema.Entity, schema.Value, schema.TS, q.getTableName(schema.SourceTable))
+	} else {
+		query = fmt.Sprintf("CREATE VIEW `%s` AS SELECT %s as entity, %s as value, to_timestamp('%s', 'YYYY-DD-MM HH24:MI:SS +0000 UTC')::TIMESTAMP as ts FROM `%s`", q.getTableName(tableName),
```

Comment:
Is this assuming the resource doesn't have a timestamp column itself?

Reply:
I'm specifically confused about teh time.UnixMilli(0).UTC() parameter

Reply:
This is something I put in the sql offlinestore. I believe it was to make the resource tables schemas consistent to make creating the training set simpler by always having entity, value, time. If the source table doesnt have its own timestamp, it replaces it with time.UnixMilli(0).UTC() because any values earlier than that breaks some databases like snowflake.